ops/model.py

```diff
@@ -797,6 +961,313 @@ def __init__(self, name: str, address_info: '_AddressDict'):
         # TODO: expose a hostname/canonical name for the address here, see LP: #1864086.
 
 
+class _Secret:
+    """Secret object.
+
+    Not meant to be instantiated directly by charm code.
+    Typically, you should obtain a Secret instance by creating one via your
+    Unit or your Application, or get one (typically one you don't own) from your Model.
+    """
+    secret_key_re = re.compile(r"^([a-z](?:-?[a-z0-9]){2,})$")
+
+    def __init__(self, backend: '_ModelBackend',
+                 id: Optional[str] = None,
+                 label: Optional[str] = None,
+                 revision: Optional[int] = None,
+                 am_owner: bool = False):
+        if not label and not id:
+            raise RuntimeError('Provide at least one of (`label`|`id`).')
+        if am_owner and revision is None:
+            # This should only happen in testing, when one instantiates Secret manually.
+            logger.error('secret {!r} (owned by this unit) has no revision!'
+                         ' Things might malfunction'.format(id))
+
+        self._backend = backend
+
+        if isinstance(id, str):
+            if not id.startswith('secret:'):  # todo: regex validation?
+                raise InvalidSecretIDError(
+                    'secret secret ID {!r} should start with `secret:`'.format(id)
+                )
+            if id.strip() != id:
+                # FIXME one of the hook tools returns some whitespace around
+                #  the secret-id. We should catch that earlier and trim it there.
+                logger.warning(
+                    'invalid characters in secret ID '
+                    '{!r}: something wrong'.format(id)
```

Comment:
I know that we don't really have a "private" log mechanism, but it's probably better to provide juju admins with a somewhat generic way to list/iterate secrets instead of logging potentially sensitive IDs.

Reply:
I removed that warning and added a secret validation mechanism that should only be hit if juju returns unexpected secret values.
Secret IDs are, to my knowledge, not 'secret' themselves. They are meant to be shared via e.g. relation data so that's as public as it gets in jujuland.
As for the juju admins, they can do `juju list-secrets`